sycl/include/CL/sycl/types.hpp

```diff
@@ -908,14 +993,16 @@ class SwizzleOp {
       1 != IdxNum && SwizzleOp::getNumElements() == IdxNum, T>::type;
 
   template <typename T>
-  using EnableIfScalarType =
-      typename std::enable_if<std::is_convertible<DataT, T>::value &&
-                              std::is_fundamental<T>::value>::type;
+  using EnableIfScalarType = typename std::enable_if<
+      std::is_convertible<DataT, T>::value &&
+      (std::is_fundamental<T>::value ||
+       std::is_same<typename std::remove_const<T>::type, half>::value)>::type;
 
   template <typename T>
-  using EnableIfNoScalarType =
-      typename std::enable_if<!std::is_convertible<DataT, T>::value ||
-                              !std::is_fundamental<T>::value>::type;
+  using EnableIfNoScalarType = typename std::enable_if<
+      !std::is_convertible<DataT, T>::value ||
+      !(std::is_fundamental<T>::value ||
+        std::is_same<typename std::remove_const<T>::type, half>::value)>::type;
```

Comment:
This seems déjà vu...

Also any hope to factorize more the code between scalar and vector cases?